relay-server/src/utils/rate_limits.rs

```diff
@@ -372,18 +387,53 @@ where
         let mut enforcement = Enforcement::default();
 
         if let Some(category) = summary.event_category {
-            let event_limits = (self.check)(scoping.item(category), 1)?;
-            let longest = event_limits.longest();
-            enforcement.event = CategoryLimit::new(category, 1, longest);
-
-            // Record the same reason for attachments, if there are any.
-            enforcement.attachments = CategoryLimit::new(
-                DataCategory::Attachment,
-                summary.attachment_quantity,
-                longest,
-            );
-
-            rate_limits.merge(event_limits);
+            if category == DataCategory::Transaction {
+                if summary.transaction_metrics_extracted {
+                    // Check and increment indexing quota.
+                    let index_limits = (self.check)(scoping.item(DataCategory::Transaction), 1)?;
+                    let longest = index_limits.longest();
+                    enforcement.event = CategoryLimit::new(DataCategory::Transaction, 1, longest);
+                    rate_limits.merge(index_limits);
+
+                    // Check processing quota, but DO NOT increment.
+                    let processing_limits =
+                        (self.check)(scoping.item(DataCategory::TransactionProcessed), 0)?;
+                    let longest = processing_limits.longest();
+
+                    // Record this as 1, since exactly one extraction is being denied, even
+                    // if we didn't increment the counter.  This works around a quirk of
+                    // CategoryLimit::is_active which only considers something limited if
+                    // the count is larger than 1.
+                    enforcement.extracted_metrics =
+                        CategoryLimit::new(DataCategory::TransactionProcessed, 1, longest);
+                    rate_limits.merge(processing_limits);
+                } else {
+                    let processing_limits =
+                        (self.check)(scoping.item(DataCategory::TransactionProcessed), 0)?;
+                    let longest = processing_limits.longest();
+
+                    // Reject the entire transaction event as we do not want to index it nor
+                    // extract metrics from it.
+                    enforcement.event = CategoryLimit::new(DataCategory::Transaction, 1, longest);
```

Comment:
I don't understand why the `longest` rate limit for `Transaction` is coming from `TransactionProcessed`. Couldn't it be lower?

Reply:
I'm not sure I follow your question.  This marks the event as enforced since if we don't have processing quota we also do not want to index the transaction, so we enforce both.  Since only processing quota now matters we take the rate limit (duration till when we're not accepting transactions anymore) from the processed rate limiter call.

Does that answer the question or did I get it entirely wrong?

Reply:
> if we don't have processing quota we also do not want to index the transaction, so we enforce both

That's correct, but unless I'm missing something the code does a different thing. To know the processing quota is exhausted, we check the processing quota; but to know whether the indexing quota is exhausted, shouldn't we check the indexing quota? So this line (423) currently enforces the indexed transaction based on the quota left for processed transactions, ignoring the two quotas are different. The other part of the branch follows the same approach, which means this case isn't "fixed" at a later point in the pipeline.

So I wonder why we don't perform two different checks (i.e. calling `(self.check)(...Transaction...)` for the two quotas, and only do one check. I'm not sure, however, if I've missed something again.

Reply:
This branch handles the case where metrics are not yet extracted.  This can only happen during the check_envelope stage.  At this point we check for processing quota.  If have exhausted this quota we do not want to ingest this event, at all.  So we enforce the event so that it will be removed and not further processed at all, not for indexing nor extracting metrics

We never do the second check for indexing quota since we only do that check *after* we have extracted the metrics (the branch above).  We should not check it since we will check it later anyway, and when we check it it will increment the count.